### DISCUSSION TITLE
Programmatically scrolling to a given point in a clickable text blob [WINUI 3]

### DISCUSSION THREAD

**Discussion by lucaregini** [microsoft/microsoft-ui-xaml#8308] [Q&A]
I need to have a control that supports a scrollable blob of text. The text displayed in there should be styled in 4 possible different ways depending on the specific content. I must be able to identify any "word" that the user has clicked on. At the same time I should be able to programmatically scroll to any of the words within the blob of text based on actions made on other controls in the same window. The content must be created and loaded dynamically. So far I haven't been able to find a way to achieve that in WINUI 3 apart from embedding a webview. Is it really my only option?

**Comment by MarkBrainfire** [comment]
I'll try to help you lucaregini.  What have you tried so far?

**Reply by lucaregini** [reply]
To scroll to a given point I must use ScrollView ChangeView method with requires a x and an y offset.

I am using a TextBlock inside a ScrollView since Rich Edit controls are not supported in UNO. 
The TextBlock contains HyperLinks so that I can detect when the user clicks on a portion of text. At the same time I want to navigate to a given hyperlink when some conditions occur.

From an HyperLink I can get a TextPointer trough its ContentStart property. What I am missing right now is how to translate this textpointer into the offset needed by ChangeView. 

This is where I am currently stuck

**Comment by castorix** [comment]
You can use a RichEditBox inside a ScrollViewer, like I did in this sample : https://github.com/castorix/WinUI3_SendMail
with a TextBox (_tbResponse_) that I scroll to the end on each server response for Net.Sockets
To scroll to a particular position, _ITextRange.GetRect_ can be used 

A test with _FindBoxHighlightMatches_ from WinUI 3 Gallery + scrolling to the found text ("Test" here) :

![RichEditBox](https://user-images.githubusercontent.com/22345506/226856743-8cd8efda-6a4f-4b52-acdc-38141c53d27c.gif)

**Reply by lucaregini** [reply]
I have a TextPointer not a TextRange. I cannot use a RichEditBox but rather a TextBlock. The question now is how can I get the Rect from the TextPointer?

**Reply by lucaregini** [reply]
Found it:

private void OnClick(Hyperlink sender, HyperlinkClickEventArgs args)
        {
            Hyperlink hyperLink = sender as Hyperlink;
            var startPointer = hyperLink.ContentStart;
            var rect= startPointer.GetCharacterRect(LogicalDirection.Forward);
            scrollView.ChangeView(0, rect.Y,null);

        }
